crates/runner/src/memory/mod.rs

```diff
@@ -168,13 +171,25 @@ impl Memory {
     /// Returns [`MemoryError::BaseFieldProjectionFailed`] if the value at the address
     /// cannot be projected to a base field element.
     pub fn get_data(&self, addr: M31) -> Result<M31, MemoryError> {
-        let address = addr.0 as usize;
-        let value = self.data.get(address).copied().unwrap_or_default();
-        if !value.1.is_zero() || !value.0 .1.is_zero() {
-            return Err(MemoryError::BaseFieldProjectionFailed { addr, value });
+        let locals_address = addr.0 as usize;
+        let heap_address = MAX_ADDRESS - addr.0 as usize;
+        // If the address is in the locals vector, read from there
+        if locals_address < self.locals.len() {
```

Comment:
Shouldn't there be a `if heap_address < self.heap.len() else return 0` ?
With this, it is assumed that if not in locals it is in heap.